sled-agent/zone-images/src/source_resolver.rs

```diff
@@ -112,41 +133,223 @@ impl ResolverInner {
-                // dataset on the boot zpool.
+                // Any zones not part of the RAM disk are managed via the
+                // zone manifest.
+                //
+                // XXX: we ask for the boot zpool to be passed in here. But
+                // `AllZoneImages` also caches the boot zpool. How should we
+                // reconcile the two?
                 if let Some(path) = internal_disks.boot_disk_install_dataset() {
-                    zone_image_paths.push(path);
+                    match self.zone_manifests.boot_disk_result() {
+                        Ok(result) => {
+                            match result.data.get(file_name.as_str()) {
+                                Some(result) => {
+                                    if result.is_valid() {
+                                        zone_image_paths.push(path);
+                                    } else {
+                                        // If the zone is not valid, we refuse to start
+                                        // it.
+                                        error!(
+                                            self.log,
+                                            "zone {} is not valid in the zone manifest, \
+                                             not returning it as a source",
+                                            file_name;
+                                            "error" => %result.display()
+                                        );
+                                    }
+                                }
+                                None => {
+                                    error!(
+                                        self.log,
+                                        "zone {} is not present in the boot disk zone manifest",
+                                        file_name,
+                                    );
+                                }
+                            }
+                        }
+                        Err(error) => {
+                            error!(
+                                self.log,
+                                "error parsing boot disk zone manifest, not returning \
+                                 install dataset as a source";
+                                "error" => InlineErrorChain::new(error),
+                            );
+                        }
+                    }
                 }
 
-                ZoneImageFileSource {
-                    file_name: install_dataset_file_name(zone_type),
+                Ok(ZoneImageFileSource {
+                    file_name,
                     search_paths: zone_image_paths,
-                }
+                })
             }
             OmicronZoneImageSource::Artifact { hash } => {
+                // TODO: implement mupdate override here. This will return an
+                // error if the override isn't found.
```

Comment:
I don't quite follow this `TODO` - if there's no mupdate override, doesn't that mean we want to run the artifacts by hash as requested?

Reply:
Yeah I got this wrong. I meant to say that this would return an error if there is an issue retrieving the override -- but it's not worth saying this in this comment. Fixed.